crates/task/src/dependency.rs

```diff
@@ -0,0 +1,270 @@
+use async_broadcast::{Receiver, RecvError};
+use futures::future::BoxFuture;
+use futures::stream::FuturesUnordered;
+use futures::stream::StreamExt;
+use futures::FutureExt;
+use std::future::Future;
+
+/// Type which describes the idea of waiting for a dependency to complete
+pub trait Dependency<T> {
+    /// Complete will wait until it gets some value `T` then return the value
+    fn completed(self) -> impl Future<Output = Option<T>> + Send;
```

Comment:
Personally I think we want to move away from `async_trait` and the MSRV doesn't really matter to us since we know our dependencies right now are on the latest rust version.  

I like the impl Future in the trait def and async fn in the implementation.  I believe for this implementation we have to use `impl Future` for the extra bounds ('Send') but I can't remember now.  TBH I wasn't aware that we could do `async fn` in the trait definition and that `impl Future` was the right/only way to get async traits without the `async_trait` crate

Reply:
I do think we want to check if this is a performance win for us.  I think we might actually benefit from boxing some of our bigger futures because the stack size might be really big for these and without box the mem has to be moved.  Idk if I fully understand but I remember from somewhere that if we see lots of CPU use in `std::mem::swap` we should look at heap allocating our futures

Reply:
But that's a probably not for this PR